srml/staking/src/phragmen.rs

```diff
@@ -16,199 +16,250 @@
 	who: AccountId,
 	// List of validators proposed by this nominator.
-	nominees: Vec<Vote<AccountId, Balance>>,
+	edges: Vec<Edge<AccountId, Balance>>,
 	// the stake amount proposed by the nominator as a part of the vote.
-	// Same as `nom.budget` in Phragmén reference.
-	stake: Balance,
+	budget: Balance,
 	// Incremented each time a nominee that this nominator voted for has been elected.
 	load: Perquintill,
 }
 
 // Wrapper around a nominator vote and the load of that vote.
-// Referred to as 'edge' in the Phragmén reference implementation.
-#[derive(Clone, Encode, Decode)]
+#[derive(Clone, Encode, Decode, Default)]
 #[cfg_attr(feature = "std", derive(Debug))]
-pub struct Vote<AccountId, Balance: HasCompact> {
+pub struct Edge<AccountId, Balance: HasCompact> {
 	// Account being voted for
 	who: AccountId,
 	// Load of this vote.
 	load: Perquintill,
 	// Final backing stake of this vote.
-	backing_stake: Balance
+	backing_stake: Balance,
+	// Index of the candidate stored in the 'candidates' vector
+	candidate_idx: usize,
+	// Index of the candidate stored in the 'elected_candidates' vector. Used only with equalise.
+	elected_idx: usize,
+	// Indicates if this edge is a vote for an elected candidate. Used only with equalise.
+	elected: bool,
 }
 
 /// Perform election based on Phragmén algorithm.
 ///
 /// Reference implementation: https://github.com/w3f/consensus
 ///
-/// @returns a vector of elected candidates
+/// Returns a vector of elected candidates
 pub fn elect<T: Trait + 'static, FR, FN, FV, FS>(
 		get_rounds: FR,
 		get_validators: FV,
 		get_nominators: FN,
 		stash_of: FS,
 		minimum_validator_count: usize,
-	) -> Vec<Candidate<T::AccountId, BalanceOf<T>>> where
-		FR: Fn() -> usize,
-		FV: Fn() -> Box<dyn Iterator<
-			Item =(T::AccountId, ValidatorPrefs<BalanceOf<T>>)
-		>>,
-		FN: Fn() -> Box<dyn Iterator<
-			Item =(T::AccountId, Vec<T::AccountId>)
-		>>,
-		FS: Fn(T::AccountId) -> BalanceOf<T>,
+		config: ElectionConfig<BalanceOf<T>>,
+) -> Vec<Candidate<T::AccountId, BalanceOf<T>>> where
+	FR: Fn() -> usize,
+	FV: Fn() -> Box<dyn Iterator<
+		Item =(T::AccountId, ValidatorPrefs<BalanceOf<T>>)
+	>>,
+	FN: Fn() -> Box<dyn Iterator<
+		Item =(T::AccountId, Vec<T::AccountId>)
+	>>,
+	for <'r> FS: Fn(&'r T::AccountId) -> BalanceOf<T>,
 {
 	let rounds = get_rounds();
-	let mut elected_candidates = vec![];
-	
+	let mut elected_candidates;
+
 	// 1- Pre-process candidates and place them in a container
 	let mut candidates = get_validators().map(|(who, _)| {
-		let stash_balance = stash_of(who.clone());
+		let stash_balance = stash_of(&who);
 		Candidate {
 			who,
-			approval_stake: BalanceOf::<T>::zero(),
-			score: Perquintill::zero(),
 			exposure: Exposure { total: stash_balance, own: stash_balance, others: vec![] },
+			..Default::default()
 		}
 	}).collect::<Vec<Candidate<T::AccountId, BalanceOf<T>>>>();
 
 	// Just to be used when we are below minimum validator count
 	let original_candidates = candidates.clone();
-	
+
```

Comment:
note that the line above this is wrong because candidates is an intention vector.

Reply:
suggest rename of `original_candidates` to `fallback_set` and passing a closure for lazily fetching fallback set to the `elect` function. in the context of staking this would be `|| session::Module<T>::validators()`

Reply:
Makes sense. I always hated this fallback mechanism anyhow. Should happen very rarely with self-vote.